libsolidity/interface/CompilerStack.cpp

```diff
@@ -194,7 +217,30 @@ bool CompilerStack::parse()
 				else
 					noErrors = false;
 
-				m_contracts[contract->name()].contract = contract;
+				if (m_contracts.find(contract->fullyQualifiedName()) != m_contracts.end())
+				{
+					const ContractDefinition* existingContract = m_contracts.find(contract->fullyQualifiedName())->second.contract;
+
+					if (contract != existingContract)
+					{
+						auto err = make_shared<Error>(Error::Type::DeclarationError);
```

Comment:
Okay.  It looks like what's happening is that there were two places in this function where contracts get added to the hashmap.  The first one is back around line 176 or so.  I replaced both cases of the insertion into the hashmap with a collision detector, because I naively believed there was a reason for both to exist; both of those inserts exist in the context of slightly different loop control logic.  I'm not sure what really differentiates them, though; it might be one isn't actually necessary.